newt/syscfg/restrict.go

```diff
@@ -82,6 +84,48 @@ type CfgRestriction struct {
 	Expr string
 }
 
+func (c CfgRestrictionCode) String() string {
+	for s, code := range cfgRestrictionNameCodeMap {
+		if code == c {
+			return s
+		}
+	}
+
+	return "???"
+}
+
+func parseCfgRestrictionCode(s string) (CfgRestrictionCode, error) {
+	if c, ok := cfgRestrictionNameCodeMap[s]; ok {
+		return c, nil
+	}
+
+	return 0, util.FmtNewtError("cannot parse cfg restriction code \"%s\"", s)
+}
+
+func (c CfgRestrictionCode) MarshalJSON() ([]byte, error) {
+	s := c.String()
+	j, err := json.Marshal(s)
+	if err != nil {
+		return nil, util.ChildNewtError(err)
+	}
+	return j, nil
+}
+
+func (c *CfgRestrictionCode) UnmarshalJSON(b []byte) error {
```

Comment:
These marshal / unmarshall functions look pretty much the same that the ones in `marshal.go` with a different struct being marshalled and a different string parsing functions, perhaps they could be standalone and called passing the string parsing function?

Reply:
That's a good idea!  I think we could eliminate a lot of duplication with something like this:
```
func MarshalJSONStringer(sr fmt.Stringer) ([]byte, error) {
    s := sr.String()
    j, err := json.Marshal(s)
    if err != nil {
        return nil, ChildNewtError(err)
    }

    return j, nil
}
```

(`fmt.Stringer` is an interface that supports the `String()` function).

Then the `CfgRestrictionCode` `MarshalJSON` function becomes:
```
func (c CfgRestrictionCode) MarshalJSON() ([]byte, error) {
    return util.MarshalJSONStringer(c)
}
```

Unfortunately, I don't think we can do something similar for the unmarshal functions.  The problem is that each unmarshal function needs to return a different type, and Go doesn't support generics.  Still, eliminating the duplication in all the marshal functions is great.